splicemachine/spliceengine
# DB-11580 Use Spark Dataset for trigger rows

## https://github.com/splicemachine/spliceengine/pull/5390 must merge before this PR.

## Short Description
For statement triggers referencing NEW TABLE or OLD TABLE, when executing on Spark, store the temporary trigger rows solely in an in-memory Spark Dataset instead of using temporary conglomerate.

## Long Description
DB-11302 https://github.com/splicemachine/spliceengine/pull/5190 improved performance for statement triggers on Spark.  One of the ways it did this was by remembering the source dataset of the INSERT, UPDATE or DELETE statement that fired the trigger, then using that instead of a temporary congomerate (written to HBase) for accessing NEW TABLE or OLD TABLE rows.
Example trigger:

> CREATE TRIGGER FeatureStoreTest.FeatureTable_History_UPDATE_OOO
> AFTER UPDATE ON FeatureStoreTest.FeatureTable
> REFERENCING NEW_TABLE AS NEWW OLD_TABLE AS OLDW
> FOR EACH STATEMENT
>     INSERT INTO FeatureStoreTest.FeatureTable (entity_key, last_update_ts, feature1, feature2) 
>     SELECT o.entity_key, o.last_update_ts, o.feature1, o.feature2
>     FROM OLDW o INNER JOIN NEWW n ON o.ENTITY_KEY=n.ENTITY_KEY AND 
>                             o.LAST_UPDATE_TS > n.LAST_UPDATE_TS
>     ;

However, it could not detect ahead of time special cases, such as those using nested loop join, or nested triggers, where the conglomerate approach is required for correctness due to execution details, so we still had to always create and write to the temporary conglomerate.  This PR adds detection for cases which can avoid writing the conglomerate, and disallows use of nested loop join for statement triggers with a referencing clause on spark.

Also as part of this PR, the default control execution cutoff where we switch to spark execution is increased from 1 million to 10 million rows.  It is costly to switch to Spark mid-stream because all the previous work has to be thrown away, so increasing the default will help avoid these cases, yet not be so high that we don't over-commit to control execution.

For the DB-11580 test case (available in https://splicemachine.atlassian.net/browse/DB-12032) the average spark runtime I'm seeing on a k8s cluster with 8 spark executors for an UPDATE with the 3 defined triggers is 34 seconds without DB-11580 and 21 seconds with DB-11580, a better than 35% improvement.  On control the average runtime is closer to 80 seconds.

## How to test

> create table t1 (a int, b int);
> create table t2 (a int, b int);
> // The following trigger should be disallowed:
> CREATE TRIGGER mytrig
> AFTER UPDATE OF a,b
> ON t1
> REFERENCING OLD TABLE AS OLD NEW TABLE AS NEW
> FOR EACH STATEMENT
> INSERT INTO T2 SELECT NEW.A, NEW.B from NEW --splice-properties joinStrategy=nestedloop
> , T1 --splice-properties joinStrategy=nestedloop, useSpark=true;
